cdap-proto/src/main/java/co/cask/cdap/proto/Id.java

```diff
@@ -349,4 +352,87 @@ public int hashCode() {
       return Objects.hashCode(namespace, category, name);
     }
   }
+
+
+  /**
+   * Id.Stream uniquely identifies a stream.
+   */
+  public static final class Stream {
+    private final Namespace namespace;
+    private final String streamName;
+    private final int hashCode;
```

Comment:
See how java.lang.String is doing it in a thread-safe way without using volatile/synchronize.